### PR TITLE
Add note on duplicate coefficients.

### LINKED ISSUE #234: Duplicate indices for coeffs
Gurobi, CPLEX, Xpress and GLPK do not accept duplicate indices natively when adding constraints.

I think that forcing all solvers to allow duplicate indices in constraints is an unnecessary overhead for adding constraints.

JuMP used to collect those, is it still true?

### PR DESCRIPTION
**Pull request by odow**
Closes #234

### CODE REVIEW

File: docs/src/apimanual.md [hunk 1]
```diff
@@ -893,27 +893,55 @@ Avoid storing extra copies of the problem when possible. This means that solver
...
+MOI defines a very general interface, with multiple possible ways to describe
+the same constraint. This is considered a feature, not a bug. MOI is designed to
+make it possible to experiment with alternative representations of an
+optimization problem at both the solving and modeling level. When implementing
+an interface, it is important to keep in mind that the constraints which a
+solver supports via MOI will have a near 1-to-1 correspondence with how users
+can express problems in JuMP, because JuMP does not perform automatic
```
Comment (odow) [inline docs/src/apimanual.md]: Should we really say this now that we have bridges?

File: docs/src/apimanual.md [hunk 2]
```diff
@@ -893,27 +893,55 @@ Avoid storing extra copies of the problem when possible. This means that solver
...
-
-- `ScalarAffineFunction` in `GreaterThan`, `LessThan`, or `EqualTo` with a nonzero constant in the function. Constants in the affine function should instead be moved into the parameters of the corresponding sets.
-
-- `ScalarAffineFunction` in `Nonnegative`, `Nonpositive` or `Zeros`. Alternative constraints are available by using a `VectorAffineFunction` with one output row or `ScalarAffineFunction` with `GreaterThan`, `LessThan`, or `EqualTo`.
-
-- Two `SingleVariable`-in-`LessThan` constraints applied to the same variable (similarly with `GreaterThan`). These should be interpreted as variable bounds, and each variable naturally has at most one upper or lower bound.
+One notable difference is that a variable with an upper and lower bound is
+translated into two constraints, rather than an interval. i.e.:
+ - `@variable(m, 0 <= x <= 1)` becomes `SingleVariable`-in-`LessThan` *and*
+   `SingleVariable`-in-`GreaterThan`.
```
Comment (odow) [inline docs/src/apimanual.md]: Is this true? I thought it got turned into `SingleVariable`-in-`Interval`?

File: docs/src/apimanual.md [hunk 3]
```diff
@@ -893,27 +893,55 @@ Avoid storing extra copies of the problem when possible. This means that solver
...
+
+Therefore, if a solver wrapper does not support
+`ScalarAffineFunction`-in-`LessThan` constraints, users will not be able to
+write: `@constraint(m, 2x + y <= 10)` in JuMP. With this in mind, developers
+should support all the constraint types that they want to be usable from JuMP.
```
Comment (odow) [inline docs/src/apimanual.md]: Ditto: bridges comment?

Comment (blegat) [reply docs/src/apimanual.md]: Yes, there isn't yet a bridge for that but later you won't have to implement ScalarAffineFunction-in-LessThan if you have VectorAffineFunction-in-Nonpositives or VectorAffineFunction-in-Nonnegatives or ScalarAffineFunction-in-EqualTo and SingleVariable-in-LessThan or ScalarAffineFunction-in-Interval or ScalarAffineFunction-in-GreaterThan or ...

File: docs/src/apimanual.md [hunk 4]
```diff
@@ -893,27 +893,55 @@ Avoid storing extra copies of the problem when possible. This means that solver
...
+  constraints are available by using a `VectorAffineFunction` with one output
+  row or `ScalarAffineFunction` with `GreaterThan`, `LessThan`, or `EqualTo`.
+
+- Two `SingleVariable`-in-`LessThan` constraints applied to the same variable
+  (similarly with `GreaterThan`). These should be interpreted as variable
+  bounds, and each variable naturally has at most one upper or lower bound.
+
+In addition, when creating functions such as `ScalarAffineFunction` and
+`ScalarQuadraticFunction`, JuMP will aggregate duplicate coefficients. For
+example, `2x + x` will be passed as `3x`, a `ScalarAffineFunction` with one
```
Comment (blegat) [inline docs/src/apimanual.md]: Bridges can create duplicate coefficients so solvers should still support them. They can call `MOIU.canonical(f)` to remove them.

File: docs/src/apimanual.md [hunk 5]
```diff
@@ -815,17 +815,31 @@ twice the value of the `coefficients` field in the `VectorAffineFunction` for
...
+(`ScalarAffineFunction`-in-`Interval`) could be re-formulated as two
+constraints: ``a^\top x ge l`` (`ScalarAffineFunction`-in-`GreaterThan`) and
+``a^\top x \le u`` (`ScalarAffineFunction`-in-`LessThan`). An alternative
+re-formulation is to add a dummy variable `y` with the constraints ``l \le y \le
+u`` (`SingleVariable`-in-`Interval`) and ``a^\top x - y = 0``
+(`ScalarAffineFunction`-in-`EqualTo`).
+
+To avoid each solver having to code these transformations manually,
+MathOptInterface provides *bridges*. A bridge is a small transformation from one
+constraint type  to another (potentially collection of) constraint type. Because
```
Comment (blegat) [inline docs/src/apimanual.md]: two spaces between "type" and "to"

File: docs/src/apimanual.md [hunk 6]
```diff
@@ -835,6 +849,27 @@ will additionally support `ScalarAffineFunction`-in-`Interval`.
...
+
+```jldoctest
+x = MathOptInterface.VariableIndex(1)
+term = MathOptInterface.ScalarAffineTerm(1, x)
+func = MathOptInterface.ScalarAffineFunction([term, term], 0)
+func_canon = MathOptInterface.Utilities.canonical(func)
+func_canon ≈ MathOptInterface.ScalarAffineFunction(
+    [MathOptInterface.ScalarAffineTerm(2, x)], 0)
+
+# Output
```
Comment (blegat) [inline docs/src/apimanual.md]: It should be `output` not `Output` (see Travis failure)

File: docs/src/apimanual.md [hunk 7]
```diff
@@ -911,27 +946,46 @@ Avoid storing extra copies of the problem when possible. This means that solver
...
+write: `@constraint(m, 2x + y <= 10)` in JuMP. With this in mind, developers
+should support all the constraint types that they want to be usable from JuMP.
+That said, from the perspective of JuMP, solvers can safely choose to not
+support the following constraints:
+
+- `ScalarAffineFunction` in `GreaterThan`, `LessThan`, or `EqualTo` with a
+  nonzero constant in the function. Constants in the affine function should
+  instead be moved into the parameters of the corresponding sets.
+
+- `ScalarAffineFunction` in `Nonnegative`, `Nonpositive` or `Zeros`. Alternative
```
Comment (blegat) [inline docs/src/apimanual.md]: The constraint **should be** `VectorAffineFunction` with one row or `ScalarAffineFunction` with ...
It is not an alternative. There is an error now when mixing scalar and vector functions and set

File: docs/src/apimanual.md [hunk 8]
```diff
@@ -911,27 +946,46 @@ Avoid storing extra copies of the problem when possible. This means that solver
...
-- `ScalarAffineFunction` in `Nonnegative`, `Nonpositive` or `Zeros`. Alternative constraints are available by using a `VectorAffineFunction` with one output row or `ScalarAffineFunction` with `GreaterThan`, `LessThan`, or `EqualTo`.
-
-- Two `SingleVariable`-in-`LessThan` constraints applied to the same variable (similarly with `GreaterThan`). These should be interpreted as variable bounds, and each variable naturally has at most one upper or lower bound.
+One notable difference is that a variable with an upper and lower bound is
+translated into two constraints, rather than an interval. i.e.:
+ - `@variable(m, 0 <= x <= 1)` becomes `SingleVariable`-in-`LessThan` *and*
+   `SingleVariable`-in-`GreaterThan`.
+
+Therefore, if a solver wrapper does not support
+`ScalarAffineFunction`-in-`LessThan` constraints, users will not be able to
```
Comment (blegat) [inline docs/src/apimanual.md]: This still need to be updated. If the model does support `ScalarAffineFunction`-in-`LessThan` but supports `VectorAffineFunction`-in-`Nonpositives` or `VectorAffineFunction`-in-`Nonnegatives` or ... it should still be ok

**Comment by odow** [comment]
Bump

File: docs/src/apimanual.md [hunk 9]
```diff
@@ -942,27 +977,43 @@ called `Optimizer`):
...
 
-One notable difference is that a variable with an upper and lower bound is translated into two constraints, rather than an interval. i.e.:
- - `@variable(m, 0 <= x <= 1)` becomes `SingleVariable`-in-`LessThan` *and* `SingleVariable`-in-`GreaterThan`.
+One notable difference is that a variable with an upper and lower bound is
+translated into two constraints, rather than an interval. i.e.:
+ - `@variable(m, 0 <= x <= 1)` becomes `SingleVariable`-in-`LessThan` *and*
+   `SingleVariable`-in-`GreaterThan`.
 
-Therefore, if a solver wrapper does not support `ScalarAffineFunction`-in-`LessThan` constraints, users will not be able to write: `@constraint(m, 2x + y <= 10)` in JuMP. With this in mind, developers should support all the constraint types that they want to be usable from JuMP. That said, from the perspective of JuMP, solvers can safely choose to not support the following constraints:
+With this in mind, developers should support all the constraint types that they
```
Comment (blegat) [inline docs/src/apimanual.md]: This is confusing "should support all the constraint types that they want to be usable from JuMP" is in opposition with the bridge section above.
I would rather say that it should support all the constraint types that directly match one structure exploited by the API, i.e. when it is *natively* supported.

Comment (blegat) [reply docs/src/apimanual.md]: This was resolved in https://github.com/JuliaOpt/MathOptInterface.jl/pull/885

**Comment by blegat** [comment]
Just rebased as it clashes with #885. Good to go now that #885 resolved the concern I had in https://github.com/JuliaOpt/MathOptInterface.jl/pull/581#discussion_r244875460

File: docs/src/apimanual.md [hunk 10]
```diff
@@ -999,19 +999,29 @@ appropriate bridges for unsupported constrained variables.
...
+re-formulation is to add a dummy variable `y` with the constraints ``l \le y \le
+u`` (`SingleVariable`-in-`Interval`) and ``a^\top x - y = 0``
+(`ScalarAffineFunction`-in-`EqualTo`).
+
+To avoid each solver having to code these transformations manually,
+MathOptInterface provides *bridges*. A bridge is a small transformation from one
+constraint type to another (potentially collection of) constraint type. Because
+these bridges are included in MathOptInterface, they can be re-used by any
+optimizer. Some bridges also implement constraint modifications and constraint
+primal and duals translations.
```
Comment (mlubin) [inline docs/src/apimanual.md]: s/duals/dual/

**Pull request merged**